native/NetworkListener/MessageManager.cpp

```diff
@@ -0,0 +1,556 @@
+    vector<wstring> parserList{ L"HtmlDownload", L"Image", L"Link", L"Media", L"ParsedElement" };
+    if (std::find(std::begin(scriptList), std::end(scriptList), initiator) != std::end(scriptList))
+    {
+        return "script";
+    }
+    else if (std::find(std::begin(parserList), std::end(parserList), initiator) != std::end(parserList))
+    {
+        return "parser";
+    }
+    else
+    {
+        return "other";
+    }
+}
+
+bool StringContainsSubstring(wstring p_string, wstring p_substring) 
+{    
+    auto it = std::search(
+        p_string.begin(), p_string.end(),
+        p_substring.begin(), p_substring.end());
+    return (it != p_string.end());    
+}
+
+wstring ToLower(wstring text) 
+{
+    wstring result = text;
+    transform(result.begin(), result.end(), result.begin(), ::towlower);
+    return result;
+}
+
+String^ ParseResourceTypeFromContentType(String^ contentType) 
+{    
+    pair<wstring, String^> resourceTypes[8]
+    { 
+        // non-mapped resource types are TextTrack, XHR, Fetch, EventSource, WebSocket
+        {L"javascript", "Script"},
+        { L"css","Stylesheet" },
+        { L"image","Image" },
+        { L"audio","Media" },
+        { L"video","Media" },
+        { L"font","Font" },
+        { L"html","Document" },
+        { L"manifest","Manifest" },
+    };
+     
+    wstring contentTypeLC = ToLower(contentType->Data());
+
+    for (int i = 0; i < 8; i++)
+    {        
+        if (StringContainsSubstring(contentTypeLC, resourceTypes[i].first)) 
+        {
+            return resourceTypes[i].second;
+        }
+    }
+
+    return "Other";
+}
+
+JsonObject ^ MessageManager::GenerateRequestWilBeSentMessage(HttpDiagnosticProviderRequestSentEventArgs ^data, String^ postPayload)
+{
+    HttpRequestMessage^ message = data->Message;
+    JsonObject^ result = ref new JsonObject();    
+    InsertString(result, "method", "Network.requestWillBeSent");
+
+    JsonObject^ params = ref new JsonObject();    
+    InsertString(params, "requestId", GetNextSequenceId(IdTypes::RequestId));
+    InsertString(params, "frameId", GetNextSequenceId(IdTypes::FrameId));
+    InsertString(params, "loaderId", GetNextSequenceId(IdTypes::LoaderId));
+    InsertString(params, "documentURL", message->RequestUri->AbsoluteUri);
+
+    JsonObject^ request = ref new JsonObject();
+    InsertString(request, "url", message->RequestUri->AbsoluteUri);
+
+    InsertString(request, "method", message->Method->Method);    
+    request->Insert("headers", SerializeHeaders(message->Headers->First()));
+    if (postPayload != nullptr && message->Method->Method == "POST")
+    {
+        InsertString(request, "postData", postPayload);
+    }
+    InsertString(request, "initialPriority", "");
+
+    params->Insert("request", request);        
+                
+    auto timeInSecs = data->Timestamp.UniversalTime / (10000000);
+    InsertNumber(params,"timestamp", timeInSecs);
+    InsertNumber(params, "walltime", 0);
```

Comment:
Yes, but I am not totally convinced. The parameter is mandatory but I cannot find this info in the WinRT event that I use to fill the Chrome message. So I let 0 in the field waiting for some feedback from the Sonar team in order to see what to put there.